File: backend/routers/proactive.py
"""Internal orchestration endpoint for canonical proactive Hermes messages."""

from __future__ import annotations

import asyncio
import hmac
import json
import os
import tempfile
import time
from pathlib import Path
from urllib.parse import quote
from uuid import uuid4

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from ..hermes_dashboard import DashboardImportError, import_assistant_session
from .notifications import NotificationPayload, deliver_notification

router = APIRouter(prefix="/api/proactive", tags=["proactive"])

MAX_IDEMPOTENCY_RECORDS = 500
_orchestration_lock = asyncio.Lock()


class ProactiveMessage(BaseModel):
    request_id: str = Field(min_length=1, max_length=128, pattern=r"^[A-Za-z0-9_.:-]+$")
    title: str = Field(default="Hermes Agent", min_length=1, max_length=200)
    message: str = Field(min_length=1, max_length=65_536)


def _records_path() -> Path:
    configured = os.environ.get("PROACTIVE_REQUESTS_FILE")
    if configured:
        return Path(configured).expanduser()
    data_dir = Path(os.environ.get("HERMES_UI_DATA_DIR", "/tmp/hermes-chat-ui"))
    return data_dir / "proactive_requests.json"


def _load_records() -> dict[str, dict]:
    path = _records_path()
    try:
        payload = json.loads(path.read_text())
    except (FileNotFoundError, OSError, json.JSONDecodeError):
        return {}
    if not isinstance(payload, dict):
        return {}
    return {
        key: value
        for key, value in payload.items()
        if isinstance(key, str) and isinstance(value, dict)
    }


def _save_records(records: dict[str, dict]) -> None:
    path = _records_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    ordered = sorted(
        records.items(),
        key=lambda item: float(item[1].get("stored_at", 0)),
        reverse=True,
    )[:MAX_IDEMPOTENCY_RECORDS]
    serializable = dict(ordered)
    temporary_name = ""
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", dir=path.parent, delete=False
        ) as temporary:
            json.dump(serializable, temporary, separators=(",", ":"))
            temporary_name = temporary.name
        os.replace(temporary_name, path)
    finally:
        if temporary_name:
            try:
                os.unlink(temporary_name)
            except FileNotFoundError:
                pass


def _require_internal_key(request: Request) -> None:
    expected = os.environ.get("HERMES_PUSH_API_KEY", "")
    if not expected:
        raise HTTPException(
            status_code=503,
            detail="Proactive messaging is not configured",
            headers={"Cache-Control": "no-store"},
        )
    supplied = request.headers.get("Authorization", "")
    if not hmac.compare_digest(supplied, f"Bearer {expected}"):
        raise HTTPException(status_code=401, detail="Unauthorized")


def _overall_status(persisted: bool, push_status: str) -> str:
    push_succeeded = push_status == "sent"
    if persisted and push_succeeded:
        return "complete"
    if persisted or push_status in {"sent", "partial"}:
        return "partial"
    return "failed"


@router.post("/messages")
async def create_proactive_message(
    payload: ProactiveMessage, request: Request
) -> dict:
    _require_internal_key(request)

    async with _orchestration_lock:
        records = await asyncio.to_thread(_load_records)
        existing = records.get(payload.request_id)
        if existing:
            replay = dict(existing.get("result") or {})
            replay["replayed"] = True
            return replay

        session_id = f"proactive_{int(time.time() * 1000)}_{uuid4().hex[:8]}"
        persistence_error: str | None = None
        try:
            await import_assistant_session(session_id, payload.title, payload.message)
            persisted = True
        except DashboardImportError as exc:
            persisted = False
            persistence_error = exc.code

        if persisted:
            push_body = payload.message
            push_url = f"/?session={quote(session_id, safe='')}"
        else:
            push_body = f"Conversation was not saved. {payload.message}"
            push_url = "/"

        push_result = await asyncio.to_thread(
            deliver_notification,
            NotificationPayload(
                title=payload.title,
                body=push_body[:500],
                url=push_url,
                tag=f"proactive-{payload.request_id}",
                notification_id=payload.request_id,
                session_id=session_id if persisted else None,
            ),
        )
        result = {
            "status": _overall_status(persisted, push_result["status"]),
            "session": {
                "persisted": persisted,
                **({"id": session_id} if persisted else {}),
                **({"error": persistence_error} if persistence_error else {}),
            },
            "push": push_result,
            "replayed": False,
        }
        records[payload.request_id] = {"stored_at": time.time(), "result": result}
        await asyncio.to_thread(_save_records, records)
        return result

File: server/proactive.ts
import { randomBytes } from "node:crypto";
import { z } from "zod";
import type { ServerConfig } from "./config.js";
import { DashboardImportError, importAssistantSession } from "./dashboard.js";
import { readJsonFile, writeJsonFileAtomic } from "./files.js";
import {
  deliverNotification,
  matchesBearer,
  type NotificationPayload,
} from "./notifications.js";

const MAX_IDEMPOTENCY_RECORDS = 500;
const ProactiveMessageSchema = z.object({
  request_id: z
    .string()
    .min(1)
    .max(128)
    .regex(/^[A-Za-z0-9_.:-]+$/),
  title: z.string().min(1).max(200).default("Hermes Agent"),
  message: z.string().min(1).max(65_536),
});

type ProactiveRecord = {
  stored_at: number;
  result: Record<string, unknown>;
};

const loadRecords = async (
  config: ServerConfig,
): Promise<Record<string, ProactiveRecord>> => {
  const payload = await readJsonFile<unknown>(config.proactiveRequestsFile, {});
  if (!payload || typeof payload !== "object" || Array.isArray(payload))
    return {};
  return Object.fromEntries(
    Object.entries(payload).filter(
      (entry): entry is [string, ProactiveRecord] => {
        const value = entry[1];
        if (!value || typeof value !== "object" || Array.isArray(value))
          return false;
        const record = value as Record<string, unknown>;
        return (
          typeof record.stored_at === "number" &&
          Boolean(record.result) &&
          typeof record.result === "object" &&
          !Array.isArray(record.result)
        );
      },
    ),
  );
};

export type ProactiveDependencies = {
  importSession: typeof importAssistantSession;
  push: typeof deliverNotification;
};

const defaultDependencies: ProactiveDependencies = {
  importSession: importAssistantSession,
  push: deliverNotification,
};

let orchestrationTail: Promise<void> = Promise.resolve();
const exclusive = async <T>(operation: () => Promise<T>): Promise<T> => {
  const previous = orchestrationTail;
  let release: () => void = () => undefined;
  orchestrationTail = new Promise<void>((resolve) => {
    release = resolve;
  });
  await previous;
  try {
    return await operation();
  } finally {
    release();
  }
};

const overallStatus = (persisted: boolean, pushStatus: string): string => {
  if (persisted && pushStatus === "sent") return "complete";
  if (persisted || ["sent", "partial"].includes(pushStatus)) return "partial";
  return "failed";
};

const saveRecords = async (
  config: ServerConfig,
  records: Record<string, ProactiveRecord>,
): Promise<void> => {
  const bounded = Object.fromEntries(
    Object.entries(records)
      .sort((left, right) => right[1].stored_at - left[1].stored_at)
      .slice(0, MAX_IDEMPOTENCY_RECORDS),
  );
  await writeJsonFileAtomic(config.proactiveRequestsFile, bounded);
};

export const createProactiveMessage = async (
  config: ServerConfig,
  authorization: string,
  input: unknown,
  dependencies: ProactiveDependencies = defaultDependencies,
): Promise<Response> => {
  if (!config.pushApiKey) {
    return Response.json(
      { detail: "Proactive messaging is not configured" },
      { status: 503, headers: { "Cache-Control": "no-store" } },
    );
  }
  if (!matchesBearer(authorization, config.pushApiKey)) {
    return Response.json({ detail: "Unauthorized" }, { status: 401 });
  }
  const parsed = ProactiveMessageSchema.safeParse(input);
  if (!parsed.success) {
    return Response.json(
      { detail: "Invalid proactive message" },
      { status: 422 },
    );
  }

  return exclusive(async () => {
    const records = await loadRecords(config);
    const existing: ProactiveRecord | undefined = Object.hasOwn(
      records,
      parsed.data.request_id,
    )
      ? records[parsed.data.request_id]
      : undefined;
    if (existing) {
      return Response.json({ ...existing.result, replayed: true });
    }

    const sessionId = `proactive_${Date.now()}_${randomBytes(4).toString("hex")}`;
    let persisted = false;
    let persistenceError: string | undefined;
    try {
      await dependencies.importSession(
        config,
        sessionId,
        parsed.data.title,
        parsed.data.message,
      );
      persisted = true;
    } catch (error) {
      persistenceError =
        error instanceof DashboardImportError
          ? error.code
          : "dashboard_unavailable";
    }

    const notification: NotificationPayload = {
      title: parsed.data.title,
      body: (persisted
        ? parsed.data.message
        : `Conversation was not saved. ${parsed.data.message}`
      ).slice(0, 500),
      url: persisted ? `/?session=${encodeURIComponent(sessionId)}` : "/",
      tag: `proactive-${parsed.data.request_id}`,
      notification_id: parsed.data.request_id,
      session_id: persisted ? sessionId : null,
    };
    const pushResult = await dependencies.push(config, notification);
    const result = {
      status: overallStatus(persisted, pushResult.status),
      session: {
        persisted,
        ...(persisted ? { id: sessionId } : {}),
        ...(persistenceError ? { error: persistenceError } : {}),
      },
      push: pushResult,
      replayed: false,
    };
    records[parsed.data.request_id] = {
      stored_at: Date.now() / 1000,
      result,
    };
    await saveRecords(config, records);
    return Response.json(result);
  });
};
